### DISCUSSION TITLE
parallel threads that can used in an activity function?

### DISCUSSION THREAD

**Discussion by ghost** [Azure/azure-functions-durable-extension#1821] [Q&A] [answered]
I have durable functions running. Timer job triggers an orchestrator, which triggers activity function that makes 10 parallel http calls on 10 threads and after the work is hands over to orchestrator and then orchestrator kicks in another orchestrator.
using an app service plan with 1 instance.

issue: how many threads can i run in parallel, or is there way i can control it. the graph shows the available threads as 50. Does that mean i can run even more than 10 in parallel. may be 20?
![query](https://user-images.githubusercontent.com/30216580/117210080-db7e5280-ae14-11eb-80db-b53a0d2efefc.PNG)

**Comment by cgillum** [comment, marked as answer]
I think what this chart is telling you is how many threads are active at any particular point in time. I can't see the x-axis, but I assume "minimum" means the minimum thread count over some interval (e.g. 5 minutes). I'm not sure why it would periodically drop to zero - perhaps that could be a glitch that occurs when scaling out. It's very likely that this number includes both threads created in your code and threads created by the functions runtime and any registered trigger listeners.

> how many threads can i run in parallel, or is there way i can control it

It depends on what language you're using for your function app. If it's C#/.NET, you can use as many threads as you want. The easiest way to schedule work on new threads is using the `Task.Run(...)` APIs (note: this should only be used in activity functions). Be careful, however, because threads are expensive and creating too many can cause significant performance problems.

If you're using Node.js or Python, your options are more limited in terms of how many threads you can allocate. By default, there is only one thread for the entire app. For Python, you can allocate more threads using the `PYTHON_THREADPOOL_THREAD_COUNT` app setting. More information here: https://docs.microsoft.com/en-us/azure/azure-functions/python-scale-performance-reference

**Reply by ghost** [reply]
@cgillum so if i want parallel execution of activities on each vm them should i cap the concurrent activities value to 1 in host.json?

**Reply by ghost** [reply]
@cgillum can you help me understand the following situation. I have an app service plan running(with 2 instances) and I set up 2 orchestrations(orch1 and orch2) which triggers  activity functions act1 and act2 respectively and after setting up max concurrent activity os 1.
orch1 -> act1 -> orch1 ->act1.....
orch2->act2 ->orch2-> act2.....
 why does activity 1 use resource of activity 2. ideally each should be running on one vm, but I dont see that happening. My act1 is faster than act2, which implies it is using resources of act2.

**Reply by cgillum** [reply]
@prady1100 I don't understand what you mean by "activity 1 use resources of act2". Do you mean they are running on the same VM? This is possible if act1 and act2 run at different times. For example, if act1 runs first and completes on VM1, it's perfectly fine if act2 also starts on VM1 if it's after act1 completes.

**Reply by ghost** [reply]
yeah the framing was wrong, to be precise, if I set concurrent running activities as 1 with 2 vm running and activity 1 and activity2 having almost same priority, then both will ideally be running parallelly right?

**Reply by cgillum** [reply]
Yes, that's right. If you invoke act1 and act2 at the same time and have maxActivityFunctionConcurrency set to 1, then they will run concurrently on two different VMs.

**Reply by ghost** [reply]
is there between cpu usage percentage vs thread count. I run out of threads and not able to do parallel processing, but the cpu usage percentage is 6-7 %. why is it this way.

**Reply by cgillum** [reply]
What language are you using? What makes you think you ran out of threads?

**Reply by ghost** [reply]
using c#. my activity function gets a batch 1000 signals. for each of them i make a http calls with 5 parallel threads. so all this activity process takes around 1 minute. if i increase the parallel processing to 10 - 12, i get no performance improvement.
increasing activity functions is also not increasing performance, i mean 2 activities with each processing 5 threads.
All this is in app service plan with 1 vm instance.
Any suggestions to increasing parallel processing?

**Reply by cgillum** [reply]
It's possible that the problem is not threads, but rather the HTTP calls you're making. Take a look at this blog post to learn about more about how to use HttpClient for parallel processing: https://makolyte.com/csharp-how-to-make-concurrent-requests-with-httpclient/. Also, make sure that the service you're calling is capable of handling multiple concurrent calls.

**Reply by ghost** [reply]
thanks @cgillum, will look into it. This discussion was very helpful.  BTW really like the work done by the azure durable functions team.
